### DISCUSSION TITLE
filament return 419 "Page Expired" when session driver be database

### DISCUSSION THREAD

**Discussion by mohamedAhmed97** [filamentphp/filament#10736] [Help]
I changed session driver to be from database not from file but that returned "419 Page Expired" and i cant login the system 
and it work when i remove "StartSession" middleware and that not right because it not create cookie form current session and in each page need to make login

**Comment by degodfather** [comment]
The same happened to me when I used SQLite as the database.

**Comment by fikurimax** [comment]
any updates?

**Reply by degodfather** [reply]
I changed the session driver to file and it started working.

**Reply by thelfensdrfer** [reply]
> it work when i remove "StartSession" middleware

I too had to remove the `StartSession` middleware in `bootstrap/app.php`.

**Comment by SeyamMs** [comment]
if you are using SQLite, make sure that server user (most likely www-data) has writing permission to the database file.

**Comment by rajentrivedi** [comment]
simple work around:

```
$panel->renderHook(
                'panels::body.end',
                fn(): string =>
                Blade::render('@livewire(\App\Livewire\NotificationPoller::class)') .
                    '<script>
                        document.addEventListener("livewire:initialized", () => {
                            Livewire.on("notification-received", (event) => {
                                const audio = new Audio("https://notificationsounds.com/storage/sounds/file-0b_oringz-pack-nine-07.ogg");
                                audio.play().catch(error => {
                                    console.error("Error playing notification sound:", error);
                                });
                            });
                            Livewire.hook("request", ({ fail }) => {
                                fail(async ({ status, preventDefault, retry }) => {
                                    if (status === 419) {
                                        preventDefault();
                                        
                                        const response = await fetch("/refresh-csrf", {
                                            method: "GET",
                                            headers: {
                                                "Accept": "application/json",
                                            },
                                            credentials: "same-origin",
                                        });

                                        const data = await response.json();
                                        const token = data.token;

                                        document.querySelector("meta[name=\\"csrf-token\\"]")
                                            .setAttribute("content", token);

                                        Livewire.csrfToken = token;
                                        retry();
                                    }
                                });
                            });
                            
                        });
                    </script>'
            )
```

Add route to web.php

```
Route::get('/refresh-csrf', function () {
    return response()->json(['token' => csrf_token()]);
})->name('refresh-csrf');
```

**Comment by albzour** [comment]
In my case, the issue was caused by the APP_NAME in the .env file containing non-English characters.

This happened because the default SESSION_COOKIE value is derived from APP_NAME:

```
// config/session.php

'cookie' => env(
    'SESSION_COOKIE',
    Str::snake((string) env('APP_NAME', 'laravel')).'_session'
),
```

The solution was to add the following line to the .env file:

`SESSION_COOKIE=any_english_name`

**Comment by shinya-dono** [comment]
### TL;DR: make sure your `sessions.user_id` column type matches `users.id` (both have to be ulid, etc...)


In my case,
i wanted to use ulid for my users table pk os i changed the 'users' migration to create a ulid column instead of the default bigInteger. i was using laravel default migration for users/sessions. so i had these migrations:

```php
// database/migrations/0001_01_01_000000_create_users_table.php
Schema::create(
	'users',
	function (Blueprint $table) {
		$table->ulid('id')->primary(); // <-- did you see it?

		$table->string('name');
		$table->string('email')->unique();
		$table->timestamp('email_verified_at')->nullable();
		$table->string('password');
		$table->rememberToken();
		$table->timestamps();
	},
);

// ...

Schema::create(
	'sessions',
	function (Blueprint $table) {
		$table->string('id')->primary();
		$table->foreignId('user_id')->nullable()->index(); // <-- i bet you did!
		$table->string('ip_address', 45)->nullable();
		$table->text('user_agent')->nullable();
		$table->longText('payload');
		$table->integer('last_activity')->index();
	},
);
```

in this scenario it would be impossible to create a `session` for a user since it cannot put the ulid in a bigInteger column.

> So you would receive a SQL exception in your `VerifyCsrfToken` middleware right?

nope...
laravel uses a rather weird approach to `upsert`ing data into database.

```php
// vendor/laravel/framework/src/Illuminate/Session/DatabaseSessionHandler.php

// ...

/**
 * Perform an insert operation on the session ID.
 *
 * @param  string  $sessionId
 * @param  array<string, mixed>  $payload
 * @return bool|null
 */
protected function performInsert($sessionId, $payload)
{
    try {
        return $this->getQuery()->insert(Arr::set($payload, 'id', $sessionId));
    } catch (QueryException) { // <-- this is the culprit i guess
        $this->performUpdate($sessionId, $payload);
    }
}

// ...
```

**Comment by RomainMazB** [comment]
For me it happened in local environment using sail and [ryoluo/sail-ssl](https://github.com/ryoluo/sail-ssl) because I was accessing the site as `http` instead of `https`.

Using `https` made it work.
